tsfc/driver.py

```diff
@@ -357,7 +359,26 @@ def compile_expression_dual_evaluation(expression, to_element, *,
                       index_cache={},
                       scalar_type=parameters["scalar_type"])
 
-    if all(isinstance(dual, PointEvaluation) for dual in to_element.dual_basis()):
+    # A FInAT QuadratureElement with a gem.Variable point set expression is the
+    # target element on the reference cell for dual evaluation where the points
+    # are specified at runtime. This special casing will not be necessary when
+    # FInAT dual evaluation is done - the dual evaluation method of such a
+    # FInAT QuadratureElement will create the necessary gem code.
+    from finat.tensorfiniteelement import TensorFiniteElement
+    runtime_quadrature_rule = (
+        isinstance(finat_to_element, QuadratureElement) or
+        (
+            isinstance(finat_to_element, TensorFiniteElement) and
+            isinstance(finat_to_element.base_element, QuadratureElement)
+        ) and
+        isinstance(finat_to_element._rule.point_set.expression, gem.Variable)
+    )
```

Comment:
This is such an anti-pattern... can you refactor please?

Reply:
To what? A `QuadratureElement` with `gem.Variable` expression defines doing runtime quadrature which, without changing quite a lot, requires flagging at the relevant places in the code. Might as well only create that flag once. Besides this is all a hack until this dual evaluation code gets sucked into FInAT (which is not the purpose of this PR).

Reply:
Also writing "this is an anti pattern" is about as not-instructive as it gets: please at least tell me _what_ the anti pattern is that you have observed otherwise you might as well be telling me "you suck and I don't like it and I won't tell you how to improve yourself"

Reply:
Also writing "this is an anti pattern" is about as not-instructive as it gets: please at least tell me _what_ the anti pattern is that you have observed otherwise you might as well be telling me "you suck and I don't like it and I won't tell you how to improve yourself"

Reply:
@miklos1 sorry if that was a bit intense - I am trying very hard, am keen to learn, and definitely value your input (particularly given you are no longer actively working on TSFC and FInAT!) Just please do remember here that there is an actual human at the other end reading these comments and who's (often difficult and iterative) work you are criticising in a public forum. :-)

Reply:
The pattern is, I guess, that everything that we're interpolating into `is-a` finat element. Ideally therefore once inside we shouldn't need to check what type of finat element it is, since all elements should offer the same API.

As you note, there was already a place where this is not true (below), in that the finat element comes in, but what we actually inspect is the FIAT element. Now, every FIAT element `has-a` dual, but not all duals are equivalent for the code generation path right now, so we special case the PointEvaluation-only cases.

With these new runtime quadrature elements, some finat elements now no longer satisfy the `has-a` fiat element property (at least, they have one, but it doesn't have a usable dual), so new special casing is introduced.

The bigger next step is to (as noted) refactor and provide a flat interface to dual evaluation from finat that is robust (and doesn't need special casing).

I think it makes sense to temporarily introduce this code (so that there's a testable code-path that new improvements can be judged against), with the intention of taking it out again once we have finat duals properly in place.

Reply:
[Switch-case code smell.](https://wiki.c2.com/?SwitchStatementsSmell)  The issue is that you are dispatching on class and object internals, and then you are doing things differently based on that.  Essentially you are observing the guts of a finite element because things are done at the wrong place.  If the internals of the finite element matter that much as to do what to do, then that functionality should live on the class, rather than here.

The canonical (or rather, mechanical) refactoring would be to ask the finite element whether its quadrature rule is run time, or even better, move all that functionality that depends on it being run time to the element classes.

Reply:
> Also writing "this is an anti pattern" is about as not-instructive as it gets: please at least tell me _what_ the anti pattern is that you have observed otherwise you might as well be telling me "you suck and I don't like it and I won't tell you how to improve yourself"

Sorry, since I generally don't have much time for code reviews, and tend to optimize and omit lengthy explanations unless they are proved to be necessary, such as being asked for.  This may sometimes come across as rude...

Reply:
Thanks for the input both, and for the lengthy explanations. So the conclusion seems to be that providing a FInAT dual evaluation interface will solve this (i.e. #218 ) so I will regard this conversation as resolved for now.